messaging/src/test/java/org/axonframework/messaging/MessageStreamTest.java

```diff
@@ -66,13 +81,96 @@ public abstract class MessageStreamTest<M extends Message<?>> {
      *
      * @return A random {@link Message} of type {@code M} to be used during testing.
      */
     abstract M createRandomMessage();
 
+    @Test
+    void shouldInvokeOnAvailableCallbackWhenMessagesAreAvailable() {
+        MessageStream<M> testSubject = completedTestSubject(List.of(createRandomMessage()));
+
+        AtomicBoolean invoked = new AtomicBoolean(false);
+        testSubject.onAvailable(() -> invoked.set(true));
+
+        assertTrue(invoked.get());
+    }
+
+    @Test
+    void shouldInvokeOnAvailableCallbackWhenStreamIsCompleted() {
+        MessageStream<M> testSubject = completedTestSubject(List.of());
+
+        AtomicBoolean invoked = new AtomicBoolean(false);
+        testSubject.onAvailable(() -> invoked.set(true));
+
+        assertTrue(invoked.get());
+    }
+
+    @Test
+    void shouldInvokeOnAvailableCallbackWhenCompleted() {
+        MessageStream<M> testSubject = failingTestSubject(List.of(), new RuntimeException("Oops"));
+
+        AtomicBoolean invoked = new AtomicBoolean(false);
+        testSubject.onAvailable(() -> invoked.set(true));
+
+        assertTrue(invoked.get());
+    }
+
+    @Test
+    void shouldNotInvokeOnAvailableCallbackWhenNotCompleted() {
+        MessageStream<M> testSubject = uncompletedTestSubject(List.of());
+
+        AtomicBoolean invoked = new AtomicBoolean(false);
+        testSubject.onAvailable(() -> invoked.set(true));
+
+        assertFalse(invoked.get());
+    }
+
+    @Test
+    void shouldNotInvokeOnAvailableCallbackWhenNotCompletedAndAvailableMessageIsConsumed() {
+        MessageStream<M> testSubject = uncompletedTestSubject(List.of(createRandomMessage()));
+
+        assertTrue(testSubject.next().isPresent());
+        assertFalse(testSubject.next().isPresent());
+
+        AtomicBoolean invoked = new AtomicBoolean(false);
+        testSubject.onAvailable(() -> invoked.set(true));
+
+        assertFalse(invoked.get());
+    }
+
+    @Test
+    void shouldNotInvokeOnAvailableCallbackWhenNotCompletedAndAvailableMessagesAreConsumed() {
+        MessageStream<M> testSubject = uncompletedTestSubject(List.of(createRandomMessage(), createRandomMessage()));
+
+        assertTrue(testSubject.next().isPresent());
+        assertTrue(testSubject.next().isPresent());
+        assertFalse(testSubject.next().isPresent());
+
+        AtomicBoolean invoked = new AtomicBoolean(false);
+        testSubject.onAvailable(() -> invoked.set(true));
+
+        assertFalse(invoked.get());
+    }
+
+    @Test
+    void shouldInvokeOnAvailableCallbackWhenNotCompletedAndNotAllAvailableMessagesAreConsumed() {
+        MessageStream<M> testSubject = uncompletedTestSubject(List.of(createRandomMessage(), createRandomMessage()));
+
+        assertTrue(testSubject.next().isPresent());
+
+        AtomicBoolean invoked = new AtomicBoolean(false);
+        testSubject.onAvailable(() -> invoked.set(true));
+
+        assertTrue(invoked.get());
+    }
+
     @Test
```

Comment:
I think it would be good to also have some rather plain tests for `next()`, `hasNextAvailable()`, `error()`, and `isCompleted()`. I am fine with picking those up, by the way.